File: src/dupr_client.py
"""DUPR API client for player lookups."""

import json
import sys
import time
from dataclasses import dataclass
from typing import Dict, List, Optional, Any
import requests

from .config import Config, debug_log


@dataclass
class PlayerRating:
    """Player rating information from DUPR."""
    singles: Optional[float]
    doubles: Optional[float]
    singles_verified: bool
    doubles_verified: bool


@dataclass
class DUPRPlayer:
    """Player information from DUPR API."""
    id: int
    full_name: str
    first_name: str
    last_name: str
    short_address: str
    ratings: PlayerRating
    dupr_id: str

    @property
    def profile_url(self) -> str:
        """Get the URL to the player's DUPR profile."""
        return f"https://dashboard.dupr.com/dashboard/player/{self.id}"

    @property
    def best_rating(self) -> Optional[float]:
        """Get the best available rating (prefer doubles)."""
        if self.ratings.doubles is not None:
            return self.ratings.doubles
        return self.ratings.singles


class DUPRAPIError(Exception):
    """Error from DUPR API."""
    pass


class TokenExpiredError(DUPRAPIError):
    """Auth token has expired."""
    pass


class RateLimitError(DUPRAPIError):
    """Rate limit exceeded."""
    pass


class DUPRClient:
    """Client for DUPR player search API."""

    def __init__(self, config: Config):
        self.config = config
        self._last_request_time = 0

    def _rate_limit_wait(self) -> None:
        """Wait to respect rate limiting."""
        elapsed = time.time() - self._last_request_time
        wait_time = (self.config.REQUEST_DELAY_MS / 1000) - elapsed
        if wait_time > 0:
            debug_log(f"Rate limiting: waiting {wait_time:.2f}s")
            time.sleep(wait_time)

    def _make_request(self, payload: Dict[str, Any]) -> Dict[str, Any]:
        """Make an API request with retry logic."""
        self._rate_limit_wait()

        headers = {
            'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64; rv:147.0) Gecko/20100101 Firefox/147.0',
            'Accept': 'application/json',
            'Accept-Language': 'en-US,en;q=0.9',
            'Content-Type': 'application/json; charset=utf-8',
            'Authorization': f'Bearer {self.config.token}',
            'Origin': 'https://dashboard.dupr.com',
            'Referer': 'https://dashboard.dupr.com/',
        }

        debug_log(f"API Request: {json.dumps(payload)}")

        for attempt in range(self.config.RETRY_COUNT):
            try:
                self._last_request_time = time.time()
                response = requests.post(
                    self.config.API_URL,
                    headers=headers,
                    json=payload,
                    timeout=30
                )

                if response.status_code == 401:
                    raise TokenExpiredError("Auth token expired - manual refresh required")

                if response.status_code == 429:
                    debug_log(f"Rate limited, waiting {self.config.RATE_LIMIT_WAIT_S}s")
                    time.sleep(self.config.RATE_LIMIT_WAIT_S)
                    continue

                response.raise_for_status()
                data = response.json()
                debug_log(f"API Response: {json.dumps(data)}")
                return data

            except requests.RequestException as e:
                debug_log(f"Request failed (attempt {attempt + 1}): {e}")
                if attempt < self.config.RETRY_COUNT - 1:
                    time.sleep(self.config.RETRY_DELAY_S)
                else:
                    raise DUPRAPIError(f"API request failed after {self.config.RETRY_COUNT} attempts: {e}")

        raise DUPRAPIError("Max retries exceeded")

    def search_players(
        self,
        query: str,
        location_text: Optional[str] = None,
        lat: Optional[float] = None,
        lng: Optional[float] = None
    ) -> List[DUPRPlayer]:
        """Search for players by name with optional location filter."""
        payload = {
            "limit": 10,
            "offset": 0,
            "query": query,
            "exclude": [],
            "includeUnclaimedPlayers": True,
            "filter": {
                "rating": {}
            }
        }

        if location_text and lat and lng:
            payload["filter"]["lat"] = lat
            payload["filter"]["lng"] = lng
            payload["filter"]["locationText"] = location_text

        response = self._make_request(payload)

        if response.get("status") != "SUCCESS":
            debug_log(f"Search failed: {response}")
            return []

        result = response.get("result", {})
        hits = result.get("hits", [])

        players = []
        for hit in hits:
            ratings_data = hit.get("ratings", {})

            # Parse ratings - handle "NR" as None
            singles = ratings_data.get("singles")
            if singles == "NR" or singles is None:
                singles = None
            else:
                singles = float(singles)

            doubles = ratings_data.get("doubles")
            if doubles == "NR" or doubles is None:
                doubles = None
            else:
                doubles = float(doubles)

            # Parse name into first/last
            full_name = hit.get("fullName", "")
            name_parts = full_name.split()
            first_name = name_parts[0] if name_parts else ""
            last_name = name_parts[-1] if len(name_parts) > 1 else ""

            player = DUPRPlayer(
                id=hit.get("id"),
                full_name=full_name,
                first_name=first_name,
                last_name=last_name,
                short_address=hit.get("shortAddress", ""),
                ratings=PlayerRating(
                    singles=singles,
                    doubles=doubles,
                    singles_verified=ratings_data.get("singlesVerified") != "NR",
                    doubles_verified=ratings_data.get("doublesVerified") != "NR"
                ),
                dupr_id=hit.get("duprId", "")
            )
            players.append(player)

        debug_log(f"Found {len(players)} players for query '{query}'")
        return players

File: packages/core/src/dupr-client.ts
/**
 * DUPR Client Module
 * Handles communication with DUPR API for player search
 */

import axios, { AxiosInstance } from 'axios';

// =============================================================================
// Constants
// =============================================================================

export const DUPR_API_URL = 'https://api.dupr.gg/player/v1.0/search';
export const REQUEST_DELAY_MS = 500;
export const RETRY_COUNT = 3;
export const RETRY_DELAY_MS = 2000;

// =============================================================================
// Types
// =============================================================================

export interface PlayerRating {
  singles: number | null;
  doubles: number | null;
  singlesVerified: boolean;
  doublesVerified: boolean;
}

export interface DUPRPlayer {
  id: number;
  fullName: string;
  firstName: string;
  lastName: string;
  shortAddress: string;
  ratings: PlayerRating;
  duprId: string;
  profileUrl: string;
  /** Best available rating (prefers doubles, falls back to singles) */
  bestRating: number | null;
}

export interface DUPRSearchRequest {
  query: string;
  limit?: number;
  offset?: number;
  locationText?: string;
  lat?: number;
  lng?: number;
}

export interface DUPRAPIResponse {
  status: 'SUCCESS' | 'ERROR';
  result?: {
    hits: DUPRAPIHit[];
    total: number;
  };
  error?: string;
}

export interface DUPRAPIHit {
  id: number;
  fullName: string;
  firstName?: string;
  lastName?: string;
  duprId: string;
  shortAddress: string;
  ratings: {
    singles: number | 'NR';
    doubles: number | 'NR';
    singlesVerified: boolean | 'NR';
    doublesVerified: boolean | 'NR';
  };
}

// =============================================================================
// Error Classes
// =============================================================================

export class DUPRAPIError extends Error {
  constructor(message: string, public statusCode?: number) {
    super(message);
    this.name = 'DUPRAPIError';
  }
}

export class TokenExpiredError extends DUPRAPIError {
  constructor(message = 'Authentication token expired') {
    super(message, 401);
    this.name = 'TokenExpiredError';
  }
}

export class RateLimitError extends DUPRAPIError {
  constructor(message = 'Rate limit exceeded') {
    super(message, 429);
    this.name = 'RateLimitError';
  }
}

export class PlayerNotFoundError extends DUPRAPIError {
  constructor(query: string) {
    super(`Player not found: ${query}`, 404);
    this.name = 'PlayerNotFoundError';
  }
}

// =============================================================================
// DUPRClient Class
// =============================================================================

export class DUPRClient {
  private token: string;
  private lastRequestTime: number = 0;
  private axiosInstance: AxiosInstance;
  private debug: boolean;

  constructor(token: string, debug = true) {
    this.token = token;
    this.debug = debug;
    this.axiosInstance = axios.create({
      baseURL: 'https://api.dupr.gg',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
    });
  }

  private log(...args: unknown[]): void {
    if (this.debug) {
      console.log('[DUPRClient]', ...args);
    }
  }

  /**
   * Wait to respect rate limiting (500ms between requests)
   */
  private async rateLimitWait(): Promise<void> {
    const now = Date.now();
    const elapsed = now - this.lastRequestTime;
    if (elapsed < REQUEST_DELAY_MS) {
      await new Promise(resolve => setTimeout(resolve, REQUEST_DELAY_MS - elapsed));
    }
    this.lastRequestTime = Date.now();
  }

  /**
   * Make a request to the DUPR API with retry logic
   */
  private async makeRequest(payload: DUPRSearchRequest): Promise<DUPRAPIResponse> {
    await this.rateLimitWait();

    let lastError: Error | null = null;

    for (let attempt = 0; attempt < RETRY_COUNT; attempt++) {
      try {
        this.log(`API request: query="${payload.query}" location="${payload.locationText || 'none'}" attempt=${attempt + 1}`);

        const response = await this.axiosInstance.post<DUPRAPIResponse>(
          '/player/v1.0/search',
          {
            filter: {
              locationText: payload.locationText || '',
              ...(payload.lat !== undefined && payload.lng !== undefined
                ? { lat: payload.lat, lng: payload.lng }
                : {}),
            },
            limit: payload.limit || 20,
            offset: payload.offset || 0,
            query: payload.query,
          }
        );

        const hitCount = response.data.result?.hits?.length ?? 0;
        this.log(`API response: status=${response.data.status} hits=${hitCount}`);

        return response.data;
      } catch (error) {
        lastError = error as Error;

        if (axios.isAxiosError(error)) {
          const status = error.response?.status;
          const message = error.response?.data?.message || error.message;
          this.log(`API error: status=${status} message="${message}" attempt=${attempt + 1}`);

          // Don't retry auth errors
          if (status === 401) {
            this.log('Token expired or invalid - throwing TokenExpiredError');
            throw new TokenExpiredError();
          }

          // Rate limit - wait longer and retry
          if (status === 429) {
            if (attempt < RETRY_COUNT - 1) {
              this.log(`Rate limited, waiting ${RETRY_DELAY_MS}ms before retry`);
              await new Promise(resolve => setTimeout(resolve, RETRY_DELAY_MS));
              continue;
            }
            throw new RateLimitError();
          }

          // Other errors - retry with backoff
          if (attempt < RETRY_COUNT - 1) {
            const delay = RETRY_DELAY_MS * (attempt + 1);
            this.log(`Retrying in ${delay}ms...`);
            await new Promise(resolve => setTimeout(resolve, delay));
            continue;
          }
        }

        this.log(`API request failed after ${RETRY_COUNT} attempts: ${lastError?.message}`);
        throw new DUPRAPIError(
          lastError?.message || 'Unknown error',
          axios.isAxiosError(error) ? error.response?.status : undefined
        );
      }
    }

    throw new DUPRAPIError(lastError?.message || 'Max retries exceeded');
  }

  /**
   * Search for players by name
   * @param query Player name to search for
   * @param locationText Optional location text (e.g., "Alberta, Canada")
   * @param lat Optional latitude for location filtering
   * @param lng Optional longitude for location filtering
   * @returns Array of matching players
   */
  async searchPlayers(
    query: string,
    locationText?: string,
    lat?: number,
    lng?: number
  ): Promise<DUPRPlayer[]> {
    const response = await this.makeRequest({
      query,
      locationText,
      lat,
      lng,
      limit: 20,
      offset: 0,
    });

    if (response.status !== 'SUCCESS' || !response.result) {
      return [];
    }

    return response.result.hits.map(apiHitToPlayer);
  }

  /**
   * Update the authentication token
   */
  setToken(token: string): void {
    this.token = token;
    this.axiosInstance.defaults.headers['Authorization'] = `Bearer ${token}`;
  }
}

// =============================================================================
// Helper Functions
// =============================================================================

/**
 * Parse rating value from API response (can be number, string, or 'NR')
 */
function parseRating(value: number | string | null | undefined): number | null {
  if (value === null || value === undefined || value === 'NR') {
    return null;
  }
  if (typeof value === 'number') {
    return value;
  }
  if (typeof value === 'string') {
    const parsed = parseFloat(value);
    return isNaN(parsed) ? null : parsed;
  }
  return null;
}

/**
 * Convert API hit to DUPRPlayer with computed bestRating
 */
export function apiHitToPlayer(hit: DUPRAPIHit): DUPRPlayer {
  const singlesRating = parseRating(hit.ratings.singles);
  const doublesRating = parseRating(hit.ratings.doubles);

  // Prefer doubles rating, fall back to singles
  const bestRating = doublesRating ?? singlesRating;

  return {
    id: hit.id,
    fullName: hit.fullName,
    firstName: hit.firstName ?? '',
    lastName: hit.lastName ?? '',
    shortAddress: hit.shortAddress,
    duprId: hit.duprId,
    profileUrl: `https://dashboard.dupr.com/dashboard/player/${hit.id}`,
    ratings: {
      singles: singlesRating,
      doubles: doublesRating,
      singlesVerified: hit.ratings.singlesVerified === true,
      doublesVerified: hit.ratings.doublesVerified === true,
    },
    bestRating,
  };
}
